### PR TITLE
make last_log_id an Option<LogId>

### LINKED ISSUE #49: Make `last_log_id` an `Option<LogId>`, or it can not express an uninitialized state
https://github.com/datafuselabs/openraft/blob/344d444c4ef84867d2ef09fb58032832b2148fcf/openraft/src/core/mod.rs#L140

### PR DESCRIPTION
**Pull request by Veeupup**
this pr fix #49 .
make struct `RaftCore`'s `last_log_id` into `Option<LogId>` to express an uninitialized state.

### CODE REVIEW

File: openraft/src/core/admin.rs [hunk 1]
```diff
@@ -217,7 +220,7 @@ impl<'a, D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>
 
         // Caveat: membership must be updated before commit check is done with the new config.
         self.core.effective_membership = EffectiveMembership {
-            log_id: self.core.last_log_id,
+            log_id: self.core.last_log_id.expect("raft core last_log_id is uninitialized"),
```
Comment (drmingdrmer) [inline openraft/src/core/admin.rs:223]: Here it's better to use the `log_id` of the returned entry in `res`.

There is another issue that updating the `effective_membership` at line 222 should be placed after checking `res`, e.g. at line 242

File: openraft/src/core/append_entries.rs [hunk 2]
```diff
@@ -26,7 +26,7 @@ impl<D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>> Ra
         &mut self,
         msg: AppendEntriesRequest<D>,
     ) -> RaftResult<AppendEntriesResponse> {
-        tracing::debug!(%self.last_log_id, %self.last_applied, msg=%msg.summary(), "handle_append_entries_request");
+        tracing::debug!({"{} {} {}", self.last_log_id.unwrap(), self.last_applied, msg.summary()}, "handle_append_entries_request");
```
Comment (drmingdrmer) [inline openraft/src/core/append_entries.rs:29]: Same as above.

**Review by drmingdrmer — COMMENTED** [review]
I'm afraid `unwrap()` is not a safe way working with `Option`.

File: openraft/src/core/admin.rs [hunk 3]
```diff
@@ -181,7 +184,7 @@ impl<'a, D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>
                             ChangeMembershipError::LearnerIsLagging {
                                 node_id: *new_node,
                                 matched: node.matched,
-                                distance: self.core.last_log_id.index.saturating_sub(node.matched.index),
+                                distance: self.core.last_log_id.unwrap().index.saturating_sub(node.matched.index),
```
[the remark this answers is not in the document]
Comment (Veeupup) [reply openraft/src/core/admin.rs:187]: in my humble opinion, here using `unwrap()` is equal to `assert!`, it is the right way to do here?

[the remark this answers is not in the document]
Comment (Veeupup) [reply openraft/src/core/admin.rs:187]: it means that we can only do this action when it is initialized.

File: openraft/src/core/admin.rs [hunk 4]
```diff
@@ -31,8 +31,8 @@ impl<'a, D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>
         &mut self,
         mut members: BTreeSet<NodeId>,
     ) -> Result<(), InitializeError> {
-        if self.core.last_log_id.index != 0 || self.core.current_term != 0 {
-            tracing::error!({self.core.last_log_id.index, self.core.current_term}, "rejecting init_with_config request as last_log_index or current_term is 0");
+        if self.core.last_log_id.unwrap().index != 0 || self.core.current_term != 0 {
+            tracing::error!({"{} {}", self.core.last_log_id.unwrap(), self.core.current_term}, "rejecting init_with_config request as last_log_index or current_term is 0");
```
[the remark this answers is not in the document]
Comment (Veeupup) [reply openraft/src/core/admin.rs:35]: `tracing` seems not allow to do expression like `unwrap()` or `unwrap_or_else()`, it needs manually format, so I did it.

File: openraft/src/core/admin.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (drmingdrmer) [reply openraft/src/core/admin.rs:187]: In theory, it is legal to change membership with an empty, uninitialized state.
Although in this codebase it hardly will.

File: openraft/src/core/admin.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (drmingdrmer) [reply openraft/src/core/admin.rs:35]: The compiler is not smart enough 🤔 

These syntaxes just compile smoothly: 
```rust
        let x = Some(1);
        tracing::debug!(x=%x.unwrap(), "message: x: Display");
        tracing::debug!(x=?x.unwrap(), "message: x: Debug");
        tracing::debug!({ x=%x.unwrap() }, "message");

```

**Pull request closed without merging**